src/app/commissioner_app.cpp

```diff
@@ -1220,47 +1196,92 @@ void CommissionerApp::MergeDataset(CommissionerDataset &aDst, const Commissioner
 #undef SET_IF_PRESENT
 }
 
-void CommissionerApp::HandlePanIdConflict(const std::string *aPeerAddr,
-                                          const ChannelMask *aChannelMask,
-                                          const uint16_t *   aPanId,
-                                          Error              aError)
+std::string CommissionerApp::OnJoinerRequest(const ByteArray &aJoinerId)
 {
-    // TODO(wgtdkp): logging
-    (void)aPeerAddr;
+    std::string pskd;
 
-    if (aError != Error::kNone)
+    auto joinerInfo = mJoiners.find({JoinerType::kMeshCoP, aJoinerId});
+    if (joinerInfo != mJoiners.end())
+    {
+        ExitNow(pskd = joinerInfo->second.mPSKd);
+    }
+
+    // Check if all joiners has been enabled.
+    joinerInfo = mJoiners.find({JoinerType::kMeshCoP, Commissioner::ComputeJoinerId(0)});
+    if (joinerInfo != mJoiners.end())
     {
-        return;
+        ExitNow(pskd = joinerInfo->second.mPSKd);
     }
 
+exit:
+    return pskd;
+}
+
+void CommissionerApp::OnJoinerConnected(const ByteArray &aJoinerId, Error aError)
+{
+    (void)aJoinerId;
+    (void)aError;
+
+    // TODO(wgtdkp): logging
+}
+
+bool CommissionerApp::OnJoinerFinalize(const ByteArray &  aJoinerId,
+                                       const std::string &aVendorName,
+                                       const std::string &aVendorModel,
+                                       const std::string &aVendorSwVersion,
+                                       const ByteArray &  aVendorStackVersion,
+                                       const std::string &aProvisioningUrl,
+                                       const ByteArray &  aVendorData)
+{
+    (void)aVendorName;
+    (void)aVendorModel;
+    (void)aVendorSwVersion;
+    (void)aVendorStackVersion;
+    (void)aVendorData;
+
+    bool accepted = false;
+
+    auto configuredJoinerInfo = GetJoinerInfo(JoinerType::kMeshCoP, aJoinerId);
+
+    // TODO(deimi): logging
+    VerifyOrExit(configuredJoinerInfo != nullptr, accepted = false);
+    VerifyOrExit(aProvisioningUrl == configuredJoinerInfo->mProvisioningUrl, accepted = false);
+
+    accepted = true;
+
+exit:
+    return accepted;
+}
+
+void CommissionerApp::OnKeepAliveResponse(Error aError)
+{
+    (void)aError;
+
+    // Dummy handler.
+}
+
+void CommissionerApp::OnPanIdConflict(const std::string &aPeerAddr, const ChannelMask &aChannelMask, uint16_t aPanId)
+{
+    (void)aPeerAddr;
     // Main thread will wait for updates to mPanIdConflicts,
     // which guarantees no concurrent access to it.
-    mPanIdConflicts[*aPanId] = *aChannelMask;
+    mPanIdConflicts[aPanId] = aChannelMask;
```

Comment:
How do we guarantee there is no race condition? I can see that `mPanIdConflicts` is accessed in CLI.

https://github.com/openthread/ot-commissioner/blob/d6ec82410e4da1263057a15ecc20bc74ea0e3f7a/src/app/cli/interpreter.cpp#L978

Reply:
The comment is stale, we have no protection against those data with current CLI implementation.

Will update the comment and add a lock for those data.

Reply:
Not in this PR?

Reply:
since this is not directly related to this PR, let's do it in a separate PR.